Add support for Openrouter models

kovyrin opened the pull request:
This PR adds support for OpenRouter models in DSPy.rb. If this is shipped, we gain access to pretty much any model in the world through a single unified interface with an ability to quickly shift between providers and models as needed.

## Usage

Here is how to use it:

```ruby
DSPy.configure do |c|
  c.lm = DSPy::LM.new(
    'openrouter/x-ai/grok-4-fast:free',
    api_key: ENV['OPENROUTER_API_KEY']
  )
end
```

Or like this for models that don't support structured output:

```ruby
DSPy.configure do |c|
  c.lm = DSPy::LM.new(
    'openrouter/deepseek/deepseek-chat-v3.1:free',
    api_key: ENV['OPENROUTER_API_KEY'],
    structured_outputs: false
  )
end
```

Finally, for those who want to use [App Attribution](https://openrouter.ai/docs/app-attribution) with OpenRouter:

```ruby
DSPy.configure do |c|
  c.lm = DSPy::LM.new(
    'openrouter/x-ai/grok-4-fast:free',
    api_key: ENV['OPENROUTER_API_KEY'],
    x_title: 'Your App Name',
    http_referrer: 'https://yourapp.com',
  )
end
```

# TODO
* [x] New adapter added
* [x] Adapter-level tests
* [x] Integration tests against the real OpenRouter API
* [x] Documentation examples for API calls using OpenRouter models

vicentereig commented:
Hey @kovyrin thanks for putting this together. I appreciate you investing your time in the project!

This is going in the right direction, usually we've done in the past a few integration tests recorded with vcr (make sure to redact tokens obvi)

Adding a new LM is usually two step process:

1. The one we we are on, where the new adapter uses Enhanced Prompting to get JSON in and out the LLM. We do non streaming clients due to the nature of DSPy's workflows and reliance on JSON responses.

2. When supported add structured outputs compatibility. I am not super familiar with open router, however it seems that we could reuse Gemini structured outputs models. We don't have to do this right away, neither in this PR or as a follow up, although recently I wrote a quick benchmark comparing Enhanced Prompting and Structured Outputs across models.

- https://openrouter.ai/docs/features/structured-outputs
- https://vicentereig.github.io/dspy.rb/blog/articles/json-modes-comparison/

Side note: I recently refactored the type system to a module that we use to generate schemas out of Sorbet signatures in ReAct Tools too.

Happy to help, thanks again for putting this together!

kovyrin commented:
@vicentereig Ok, I think I'm ready for round 2!

Added more tests including integration tests against the real API with multiple models, multiple prompts, etc. Updated docs as well.

Let me know what you think.

vicentereig commented:
@kovyrin You got it. There was a small naming conflict that I corrected. Let me know if you want to take more time on structured outputs on a separate PR, that'd make a good next 0.28.0 release!

kovyrin commented:
Great stuff! I'll look into how we can improve structured outputs support for OpenRouter models over the next couple of days as I'm playing with DSPy.rb for my project.

vicentereig commented:
@kovyrin good stuff - I found an issue with Gemini structured outputs. Planning to fix it this week and then release OpenRouter alongside the more robust SO for Gemini models.

I am getting more familiar with OpenRouter now, it seems that you could reuse a lot of what's done already. Their SO look pretty similar to Gemini's.
